### PR TITLE
slice: more methods

### PR DESCRIPTION
**Pull request by ninedraft**
And palliative tests

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,3 +1,3 @@
 module github.com/go-perf/golds
 
-go 1.17
+go 1.18
```
Comment (cristaloleg) [inline go.mod]: 🚀

File: slice.go [hunk 2]
```diff
@@ -2,64 +2,61 @@ package golds
...
 		return nil
 	case n < 0:
 		panic("golds.Repeat: negative Repeat count")
 	}
-	var s = make(slice[E], n)
-	// TODO: use copy to fill slice for better performance
-	for i := range s {
-		s[i] = v
-	}
+	var s = make(Slice[E], n)
```
Comment (cristaloleg) [inline slice.go:21]: pull into `default` ?

File: slice.go [hunk 3]
```diff
@@ -2,64 +2,61 @@ package golds
...
-	switch{
+func SliceIterFn[E any](n int, fn func(int) E) Slice[E] {
+	switch {
 	case n == 0:
 		return nil
 	case n < 0:
 		panic("golds.Repeat: negative Repeat count")
 	}
-	var s = make(slice[E], n)
+	var s = make(Slice[E], n)
```
Comment (cristaloleg) [inline slice.go:34]: same here

File: slice.go [hunk 4]
```diff
@@ -2,64 +2,61 @@ package golds
...
 // Cap returns capacity of underlying slice.
-func(s slice[E]) Cap() int { return cap(s) }
+func (s Slice[E]) Cap() int { return cap(s) }
 
 // Index returns element by the index.
 // If index is negative, then returns element len-index.
 // Panics if index is out of [-len, len] range.
-func(s slice[E]) Index(i int) E {
+func (s Slice[E]) Index(i int) E {
 	if i < 0 {
```
Comment (cristaloleg) [inline slice.go:51]: what if `i` is less then `-2*|s|` ? :)

File: slice.go [hunk 5]
```diff
@@ -70,58 +67,69 @@ func(s slice[E]) Swap(i, j int) {
...
+func (s Slice[E]) FilterInPlace(fn func(v E) bool) {
+	var i int
+	for _, v := range s {
+		if fn(v) {
+			s[i] = v
+			i++
+		}
+	}
+	var empty E
+	s[i:].Fill(empty)
```
Comment (cristaloleg) [inline slice.go:101]: nice

File: slice.go [hunk 6]
```diff
@@ -131,4 +139,142 @@ func(s *slice[E]) Pop() (E, bool) {
...
\ No newline at end of file
+}
+
+func (s Slice[E]) Copy() Slice[E] {
+	var cp = make(Slice[E], s.Len())
+	copy(cp, s)
+	return cp
+}
+
+func (s Slice[E]) CopyWith(fn func(E) E) Slice[E] {
```
Comment (cristaloleg) [inline slice.go:150]: What is the idea behind `fn` ? I mean when and why we need `CopyWith`

File: slice.go [hunk 7]
```diff
@@ -131,4 +139,142 @@ func(s *slice[E]) Pop() (E, bool) {
...
+
+// Reverse reverses the items of slice in place.
+func (slice Slice[E]) Reverse() {
+	Reverse(slice)
+}
+
+// Reverse reverses the items of slice in place.
+func Reverse[E any](slice []E) {
+	var n = len(slice)
+	for i := 0; i < n/2; i++ {
```
Comment (cristaloleg) [inline slice.go:276]: just `for i, j := 0, len(slice)-1; i < j; i, j = i+1, j-1 {` ?

**Review by cristaloleg — APPROVED** [review]

File: slice.go (hunk 6 above)
Comment (ninedraft) [reply slice.go:150]: To use user defined deep copy

File: slice.go (hunk 4 above)
Comment (ninedraft) [reply slice.go:51]: then panek :smiling_imp:

**Review by cristaloleg — APPROVED** [review]

**Pull request merged**